doc/developer/design/20231117_copy_to_s3.md

```diff
@@ -0,0 +1,279 @@
+
+#### Other Alternative: Storage cluster
+The other alternative would have been to actually make use of a Storage cluster. If we think
+of `COPY ... TO S3` as a one time sink, then it might also make sense to have it in the Storage
+cluster. Especially if the initial scope does not include `SELECT` queries which
+would have required spinning up compute dataflows.
+Reasons to not do so,
+- As mentioned above, we will eventually support `SELECT` queries in `COPY ... TO S3`. The
+only reason it would have made sense to do this in storage would be to do as a quick stop gap
+if the effort is drastically less. It does not appear to be so.
+- Unlike Compute, Storage does not have a concept of running temporary/one-off dataflows.
+Everything which runs, like sources or sinks, do so continuously once scheduled till they are
+dropped.
+
+### Output file format
+We should start with a single file format and add more if there's more user interest. Currently,
+there has been request for parquet but CSV support would be easiest to build.
+
+Postgres supports `COPY ... TO` with a csv file and we should mimic the file structure for our
+export as well.
+
+Eventually we'll need to add some additional formatting options to the SQL like
+`DELIMITER` (default will be comma), `QUOTE` (default is double quotes) etc. In the initial
+version we should stick to the default values as per the
+[Postgres COPY command](https://www.postgresql.org/docs/current/sql-copy.html).
+
+### Bounded Memory
+I had wrongly assumed that with reading directly from persist we could use backpressure similar to what we do in upsert. As @guswynn correctly pointed out, that wouldn't work due to the following:
+* For upsert, we read configured amount of data from persist, but we eventually consolidate in rocksdb (or in memory if disk is not enabled) which maintains the upsert state.
+* In compute, we may be able to fetch data in chunks from persist, but it will still need to be
+consolidated for the entire shard and this will have to happen in memory.
+
+One possibility could be that we actually use disk (with rocksdb) in Compute to keep state similar
+to upsert and consolidate the data there before writing the snapshot to S3. This seems complicated
+and I don't think it will be worth the effort. Also, given that we might
+have arrangements spilling to disk in the future, we'll probably get this for free.
+
+TODO: Check on the status of consolidation on read from persist, if it can be used here.
+
+Since backpressure is not feasible right off the bat when reading from materialized view as I had
+assumed, there's no reason to limit the scope to only support materialized views/tables and
+we can support arbitrary SELECT queries as well.
+
+### Writing to S3
+While writing to S3, we'll need to split the data into multiple files so that each file is
+under the configured MAX FILE SIZE. In the initial version, we can have a single timely worker
```

Comment:
Compute tries pretty hard to evenly balance work between workers, so if this ends up being run in compute, I'm not sure this is the place to cut scope for a v1 :). (I also don't think it's terribly hard to make it multi worker? Each worker could be doing the logic you mention below but pointed at some `<prefix>/<worker_idx>/part-<idx>.csv`)

Reply:
Instead of the path, maybe we can add the worker_id to the filename itself, something like `<prefix>/part-<worker_idx>-<idx>.csv`. I am not sure how would this affect ordering though (there's an open question around supporting ORDER BY).

The reason being, eventually we want the data exported to S3 be used in a data warehouse like Snowflake most likely via external tables. In that scenario, all the data files should be present at the given prefix. We could potentially make it part of the path like a partition `<prefix>/worker_id=<worker_idx>/part-<idx>.csv`, and import `<prefix>` as the table, but this adds the worker ID partition column unnecessarily. There might be other ways to read sub folders, but they won't work right out of the box.

Reply:
While copying to S3, if the given path is not empty, the command is going to fail (similar to Snowflake). Also, we are going to support scalar expressions in s3 path. I would imagine for most of the use cases the path is going to be something like `'s3://some/path' + mz_now_or_some_date_time_suffix`. So, I would imagine there's very less chance of conflict happening but it's definitely not impossible. There's a window between the COPY command and when the first file is written. Hmm...maybe we should write an in progress placeholder?

But I do like the idea of writing a manifest file (we used to use versions of this widely at Stripe for various consistency checks). Snowflake does not, but Redshift does have a concept of manifest file to load data from S3 https://docs.aws.amazon.com/redshift/latest/dg/loading-data-files-using-manifest.html.

I would prefer if the folder structure like below, i.e. not nested under batches so that the data is exactly under the user specified path (and they can use it to load it into data warehouses predictably).
```
<prefix>
│── aaaa.csv
│── bbbb.csv
│── cccc.csv
└── progress_[0]-[<as_of>+1].json
```
I will have to check though if Snowflake can successfully ignore separate manifest files.

Reply:
Snowflake does support `PATTERN = 'regex_pattern'` in their `CREATE EXTERNAL TABLE` which can be used to only include files of a certain pattern.

Reply:
I can see pros and cons to each. No strong feelings myself. Is this the only open question? If so we could potentially mark it down as an open question that we'll resolve later during the implementation.

Reply:
@benesch Yes, this is the only one. Sounds good, adding it to open questions.

Reply:
Added to open questions